src/Bicep.Core/TypeSystem/DeployTimeConstantValidator.cs

```diff
@@ -0,0 +1,87 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Bicep.Core.Diagnostics;
+using Bicep.Core.Extensions;
+using Bicep.Core.Semantics;
+using Bicep.Core.Syntax;
+
+namespace Bicep.Core.TypeSystem
+{
+    public static class DeployTimeConstantValidator
+    {
+        public static void Validate(SemanticModel semanticModel, IDiagnosticWriter diagnosticWriter)
+        {
+            // Collect all sytnaxes that require DTCs (a.k.a. DTC containers).
+            var containers = DeployTimeConstantContainerVisitor.CollectDeployTimeConstantContainers(semanticModel);
+
+            foreach (var container in containers)
+            {
+                // Only visit child nodes of the DTC container to avoid flagging the DTC container itself.
+                foreach (var childContainer in GetChildrenOfDeployTimeConstantContainer(semanticModel, container))
+                {
+                    // Validate property accesses, array accesses, resource accesses and function calls.
+                    new DeployTimeConstantDirectViolationVisitor(container, semanticModel, diagnosticWriter)
+                        .Visit(childContainer);
+
+                    // Validate variable dependencies.
+                    foreach (var variableDependency in VariableDependencyVisitor.GetVariableDependencies(semanticModel, childContainer))
+                    {
+                        new DeployTimeConstantIndirectViolationVisitor(container, variableDependency, semanticModel, diagnosticWriter)
+                            .Visit(variableDependency);
+                    }
+                }
+            }
+        }
+
+        private static IEnumerable<SyntaxBase> GetChildrenOfDeployTimeConstantContainer(SemanticModel semanticModel, SyntaxBase deployTimeConstantContainer) => deployTimeConstantContainer switch
+        {
+            ObjectPropertySyntax objectPropertySyntax => objectPropertySyntax.Value.AsEnumerable(),
+            IfConditionSyntax ifConditionSyntax => ifConditionSyntax.ConditionExpression.AsEnumerable(),
```

Comment:
Do we need to look at the if body here?

Reply:
No - the properties of the if body will be collected by the DeployTimeContainerVisitor as separate nodes if they accepts DTC values.